### PR TITLE
Improve sphinx reStructuredText parsing

### PR DESCRIPTION
**Pull request by krassowski**
This should largely fix https://github.com/pappasam/jedi-language-server/issues/172.

### CODE REVIEW

File: tests/test_rst.py [hunk 1]
```diff
@@ -624,6 +624,28 @@ def func(): pass
 """
 
 
+# https://www.sphinx-doc.org/en/master/usage/restructuredtext/domains.html#info-field-lists
+SPHINX_SIGNATURE = """
```
Comment (flaviut) [inline tests/test_rst.py:628]: Is there code that handles any code blocks before this? Otherwise something like 

```
.. code-block:: python
    def foo():
        """:param  str message_body: blah blah"""
```

could cause problems. I'll admit it's an edge case that would likely only come up in docstring parsing code.

**Comment by flaviut** [comment]
Thanks for the link! I like the approach you've taken, [I've used a similar approach in the past to convert markdown to HTML](https://github.com/flaviut/stupid-markdown-parser/blob/master/markdown.nim).

File: tests/test_rst.py (hunk 1 above)
Comment (flaviut) [reply tests/test_rst.py:628]: Oh, one other thing that comes to mind! If the programmer put these directives in a random order for some reason, it would be nice if all directives of a certain type were grouped together. Also definitely a bit of an edge case & not required, but there's plenty of less-than-finely crafted code out there.

Comment (krassowski) [reply tests/test_rst.py:628]: No, it will not cause problems. Code blocks are properly parsed out at an earlier stage. I added a test case in 1771df2f1472266de9b8fb04c87111d1baea11b5 to demonstrate this.

Comment (krassowski) [reply tests/test_rst.py:628]: I agree about the order thing. That would require a proper parser though so we cannot just use regexp here. It is definitely in scope, but I don't have more time this weekend. Would you mind opening an issue so we can track this as a future improvement?

Comment (flaviut) [reply tests/test_rst.py:628]: resolved with https://github.com/python-lsp/docstring-to-markdown/issues/14
